[MRG] Bug fix for sklearn.metrics.cluster.normalized_mutual_info_score with sparse input

scouvreur opened the pull request:
#### Reference Issues/PRs

Fix for bug reported in #13836 by @smsaladi
See also related issues #12940, #13847

#### What does this implement/fix? Explain your changes.

- [x] Added assertion test for when input to `normalized_mutual_info_score()` is sparse to check the returned score is between 0 and 1.
- [x] Control structure to return fixed value in the case that `normalizer` < `eps`.

scouvreur committed:
Added test for nmi with sparse input

Added assertion test for when input to
normalized_mutual_info_score() is sparse
to check the returned score is between 0
and 1.

On branch nmi_bug_fix
Changes to be committed:
	modified:   test_supervised.py

jeremiedbb commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -140,6 +140,54 @@ def test_beta_parameter():
+    # test that on input nmi returns
+    # a score which is between 0 and 1
+    x = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
```
You can write it in a more compact way which I think will make it more readable
```python
x = [0] * <length you want>
y = [1] + x[1:]
```

scouvreur committed:
Simplified syntax for sparse test input

Simplified sparse input creation and
fixed whitespace error in CircleCI.

On branch nmi_bug_fix
Your branch is up to date with 'origin/nmi_bug_fix'.

Changes to be committed:
	modified:   test_supervised.py

scouvreur committed:
Control structure for normalizer > eps

Added control structure for when normalizer >
numpy epsilon for 64-bit float datatype.

On branch nmi_bug_fix
Your branch is up to date with 'origin/nmi_bug_fix'.

Changes to be committed:
	modified:   ../supervised.py

ogrisel commented on sklearn/metrics/cluster/supervised.py:
```diff
@@ -870,6 +870,10 @@ def normalized_mutual_info_score(labels_true, labels_pred,
+
+    # Handling sparse input
+    if normalizer < np.finfo('float64').eps:
+        return 0
```
The coverage report seems to indicate that no test can trigger this case.

scouvreur replied:
Ah apologies @ogrisel I realise this is not such a good approach as anyways :

```normalizer = max(normalizer, np.finfo('float64').eps)```

the minimum for `normalizer` cannot be less than epsilon.

scouvreur replied:
So I don't think this control structure would ever be triggered

scouvreur committed:
Added control structure to handle

Test stub created for when normalizer has
very small value, yet to find a test case
for h_true and h_pred which would return a
normalizer with a very small value.

On branch nmi_bug_fix
Your branch is up to date with 'origin/nmi_bug_fix'.

Changes to be committed:
	modified:   supervised.py
	modified:   tests/test_supervised.py

scouvreur commented:
Sorry guys - could I get some help on creating some test cases which might trigger `normalizer < np.finfo('float64').eps` ?

jnothman commented:
> Sorry guys - could I get some help on creating some test cases which might trigger normalizer < np.finfo('float64').eps ?

Haven't you said that you don't think this case can be triggered?

jnothman reviewed (commented):
Maybe I just misunderstood you.

Anyway, we want an entropy of 0? That should happen with an entirely peaked distribution?

scouvreur committed:
Added tests for nmi

Added tests for nmi to assert value returned
between 0 and 1, and greater than eps for
64 bit datatype.

scouvreur committed:
Merge remote-tracking branch 'upstream/master'

scouvreur commented:
Would this approximate a peaked distribution ?

```python
# sample peaked distribution
x = [0] * 20
y = [1] * 20
```

jnothman commented:
Not sure what x and y are there

scouvreur commented:
Hey @jnothman, in this case `x` would be the random variable and `y` the response variable.

cmarmo commented:
I believe the issue has been solved in the current version (see [13836-comment](https://github.com/scikit-learn/scikit-learn/issues/13836#issuecomment-1051221208)).
